samples/nrf9160/lwm2m_client/src/ui/ui_led.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <drivers/pwm.h>
+#include <drivers/gpio.h>
+
+#include "ui_led.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ui_led, CONFIG_UI_LOG_LEVEL);
+
+#define LED_PWM_NODE				DT_NODELABEL(pwm0)
+#define LED_PWM_CHANNEL(channel)	DT_PROP(LED_PWM_NODE, ch##channel##_pin)
+#define LED_PWM_FLAGS				DT_PWMS_FLAGS(LED_PWM_NODE)
+#define LED_PWM_DEV_LABEL			DT_LABEL(LED_PWM_NODE)
+
+#define PWM_PERIOD_USEC				(USEC_PER_SEC / 100U)
+
+#define LED_GPIO_PARENT_NODE		DT_PATH(leds)
+#define LED_GPIO_NODE(number)		DT_CHILD(LED_GPIO_PARENT_NODE, led_##number)
+#define LED_GPIO_PIN(number)		DT_GPIO_PIN(LED_GPIO_NODE(number), gpios)
+#define LED_GPIO_FLAGS				DT_GPIO_FLAGS(LED_GPIO_NODE(1), gpios)
+#define LED_GPIO_DEV_LABEL			DT_GPIO_LABEL(LED_GPIO_NODE(1), gpios)
+
+static const struct device *led_dev;
+static uint32_t pulse_width[NUM_LEDS];
+static bool state[NUM_LEDS];
+
+static int get_pwm_channel(uint8_t led_num, uint32_t *pwm_channel)
+{
+	switch (led_num) {
+	case 0:
+		*pwm_channel = LED_PWM_CHANNEL(0);
+		break;
+#if defined(CONFIG_BOARD_THINGY91_NRF9160NS)
+	case 1:
+		*pwm_channel = LED_PWM_CHANNEL(1);
+		break;
+	case 2:
+		*pwm_channel = LED_PWM_CHANNEL(2);
+		break;
+#endif
+	default:
+		LOG_ERR("PWM channel %u not supported (%d)", led_num, -ENOTSUP);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int ui_led_pwm_on_off(uint8_t led_num, bool new_state)
+{
+	int ret;
+	uint32_t pwm_channel;
+
+	state[led_num] = new_state;
+
+	ret = get_pwm_channel(led_num, &pwm_channel);
+	if (ret) {
+		return ret;
+	}
+	ret = pwm_pin_set_usec(led_dev, pwm_channel, PWM_PERIOD_USEC,
+			pulse_width[led_num] * new_state, LED_PWM_FLAGS);
+	if (ret) {
+		LOG_ERR("Set LED PWM pin %u failed  (%d)", led_num, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static uint32_t calculate_pulse_width(uint8_t led_intensity)
+{
+	return PWM_PERIOD_USEC * led_intensity / UINT8_MAX;
+}
+
+int ui_led_pwm_set_intensity(uint8_t led_num, uint8_t led_intensity)
+{
+	int ret;
+	uint32_t pwm_channel;
+
+	pulse_width[led_num] = calculate_pulse_width(led_intensity);
+
+	ret = get_pwm_channel(led_num, &pwm_channel);
+	if (ret) {
+		return ret;
+	}
+	ret = pwm_pin_set_usec(led_dev, pwm_channel, PWM_PERIOD_USEC,
+				pulse_width[led_num] * state[led_num], LED_PWM_FLAGS);
+	if (ret) {
+		LOG_ERR("Set LED PWM pin %u failed (%d)", led_num, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+int ui_led_pwm_init(void)
+{
+	led_dev = device_get_binding(LED_PWM_DEV_LABEL);
+	if (!led_dev) {
+		LOG_DBG("Could not bind to LED PWM device (%d)", -ENODEV);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int get_gpio_pin(uint8_t led_num, gpio_pin_t *gpio_pin)
+{
+	uint8_t led_id;
+
+	/* LEDs are named led_1 .. led_3 in Thingy dts, but led_0 .. led_4 in DK dts */
+	if (IS_ENABLED(CONFIG_BOARD_THINGY91_NRF9160NS)) {
```

Comment:
Yes, I've tried to use the IS_ENABLED() macro when possible, as I think it is easier to read and allows error generation even when the macro being checked is not defined. But using both IS_ENABLED() and #ifdef in the same function might be confusing?

Reply:
Yeah, sticking to one way of doing things would be good. I don't have have a preference which one to use though.

Reply:
Only problem is they don't do the exact same thing. IS_ENABLED allows for compiler errors, while #ifdefs can be used to e.g. include/exclude switch cases (as done just below the excerpt above). Also #ifdefs must be used if the code inside depends on some macro that might not be defined otherwise (which is the case in sensor_module.c). I can revert all IS_ENABLED back to #ifdefs, but I think the readability might suffer.

Reply:
Generally, it's preferred to use runtime `IS_ENABLED()` check if possible, to maximize code coverage during build, it's easier to spot issues with different configurations during a single build. It's not always possible thoguh for the reason explain above, but IMO we should stick to runtime checks wherever applicable.